pyirf/interpolation.py

```diff
@@ -0,0 +1,270 @@
+        Interpolation method
+
+    Returns
+    -------
+    aeff_interp: astropy.units.Quantity[area]
+        Interpolated Effective area array with shape (n_energy_bins, n_fov_offset_bins)
+    """
+
+    _, n_fov_offset_bins, n_energy_bins = aeff_all.shape
+
+    # get rid of units
+    aeff_all = aeff_all.to('m2').value
+    min_effective_area = min_effective_area.to('m2').value
+
+    # remove zeros and log it
+    aeff_all[aeff_all < min_effective_area] = min_effective_area
+    aeff_all = np.log(aeff_all)
+
+    # interpolation
+    aeff_interp = np.empty((n_energy_bins, n_fov_offset_bins))
+    for i_th in range(n_fov_offset_bins):
+        for i_en in range(n_energy_bins):
+            aeff_interp[i_en, i_th] = griddata(pars_all, aeff_all[:, i_th, i_en], interp_pars, method=method)
+
+    # exp it and set to zero too low values
+    aeff_interp = np.exp(aeff_interp)
+    aeff_interp[aeff_interp < min_effective_area * 1.1] = 0  # 1.1 to correct for numerical uncertainty and interpolation
+    return aeff_interp * u.Unit('m2')
+
+
+def interpolate_dispersion_matrix(matrix_all, pars_all, interp_pars, method='linear'):
+    """
+    Takes a grid of dispersion matrixes for a bunch of different parameters
+    and interpolates it to given value of those parameters
+
+    Parameters
+    ----------
+    matrix_all: np.array of astropy.units.Quantity[area]
+        grid of effective area, of shape (n_grid_points, n_fov_offset_bins, n_migration_bins, n_energy_bins)
+    pars_all: np.array
+        list of parameters corresponding to matrix_all, of shape (n_grid_points, n_interp_dim)
+    interp_pars: np.array
+        values of parameters for which the interpolation is performed, of shape (n_interp_dim)
+    method: 'linear’, ‘nearest’, ‘cubic’
+        Interpolation method
+
+    Returns
+    -------
+    matrix_interp: astropy.units.Quantity[area]
+        Interpolated dispersion matrix 3D array with shape (n_energy_bins, n_migration_bins, n_fov_offset_bins)
+    """
+
+    _, n_fov_offset_bins, n_migration_bins, n_energy_bins = matrix_all.shape
+
+    # interpolation
+    r_th = range(n_fov_offset_bins)
+    r_en = range(n_energy_bins)
+    r_mig = range(n_migration_bins)
+    matrix_interp = np.empty((n_energy_bins, n_migration_bins, n_fov_offset_bins))
+    # TO DO this part can be optimized because nested for loops take quite some time
+    # but it is not a big problem because this has to be done only once per run
+    for i_th in r_th:
+        for i_en in r_en:
+            for i_mig in r_mig:
+                matrix_interp[i_en, i_mig, i_th] = griddata(pars_all, matrix_all[:, i_th, i_mig, i_en], interp_pars, method=method)
+
+    # now we need to renormalize along the migration axis
+    norm = np.sum(matrix_interp, axis=1, keepdims=True)
+    mig_norm = np.divide(matrix_interp, norm, out=np.zeros_like(matrix_interp), where=norm != 0)
+    return mig_norm
+
+
+def read_unit_from_HDUL(hdul, ext_name, field_name):
```

Comment:
Please just use `astropy.table.Table` to read in the IRFs, they automatically handle units.

Reply:
and metadata! So if the parameters of the bin are in the IRF fits file headers, you automatically get them in the `.meta` attribute of your table.

Reply:
thanks for the tip on this @maxnoe, I implemented your solution and thus was able to remove this function

@kosack, as you say the meta data should be then read as well, but I'm not sure how useful those will be in reality. For the moment interpolation is in cos zd and Az, but the functions can be used for the whole zoo of interpolations in NSB, PSF parameters, etc. I do not think all of this will end up in the meta data, and even if so one would still need to relate the values stored in the meta data of IRFs with the values stored in the event lists of the data. Let's come back to this discussion once we have IRF files with meta data